src/coprocessor/codec/batch/rows.rs

```diff
@@ -0,0 +1,716 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO: Remove this.
+#![allow(dead_code)]
+
+use std::convert::TryFrom;
+
+use smallvec::SmallVec;
+
+use cop_datatype::{EvalType, FieldTypeAccessor, FieldTypeFlag};
+use tipb::schema::ColumnInfo;
+
+use super::BatchColumn;
+use coprocessor::codec::mysql::Tz;
+use coprocessor::codec::{datum, Error, Result};
+
+pub struct BatchRows<E> {
+    /// Multiple interested columns. Each column is either decoded, or not decoded.
+    ///
+    /// For decoded columns, they may be in different types. If the column is in
+    /// type `LazyBatchColumn::Encoded`, it means that it is not decoded.
+    columns: Vec<LazyBatchColumn>,
+
+    /// The number of rows (either encoded or decoded) stored.
+    number_of_rows: usize,
```

Comment:
Can we infer this value from columns?

Reply:
Yes, it should be strictly equals to `columns[0].len()`. It is ok to remove this field, but I don't know if it is better than current design. Current design allows us to easily assert consistency of each column's length, but introduces extra code to sync "number_of_rows".

Reply:
I prefer a single point of truth. You can just provide a simple function called `number_of_rows()`.